extensions/m_shedding.c

```diff
@@ -0,0 +1,160 @@
+
+static void mo_shedding(struct MsgBuf *msgbuf_p, struct Client *client_p, struct Client *source_p, int parc, const char *parv[]);
+static void user_shedding_main(void *rate);
+static void user_shedding_shed(void *unused);
+
+static struct Message shedding_msgtab = {
+	"SHEDDING", 0, 0, 0, 0,
+	{mg_unreg, mg_not_oper, {mo_shedding, 3}, mg_ignore, mg_ignore, {mo_shedding, 3}}
+};
+
+mapi_clist_av1 shedding_clist[] = { &shedding_msgtab, NULL };
+
+static void
+moddeinit(void)
+{
+	rb_event_delete(user_shedding_main_ev);
+	rb_event_delete(user_shedding_shed_ev);
+}
+
+DECLARE_MODULE_AV2(shed, NULL, moddeinit, shedding_clist, NULL, NULL, NULL, NULL, shed_desc);
+
+/*
+ * mo_shedding
+ *
+ * inputs - pointer to server
+ *    - pointer to client
+ *    - parameter count
+ *    - parameter list
+ * output -
+ * side effects - user shedding is enabled or disabled
+ *
+ * SHEDDING <server> OFF - disable shedding
+ * SHEDDING <server> <approx_seconds_per_userdrop> :<reason>
+ * (parv[#] 1        2                             3)
+ *
+ */
+static void
+mo_shedding(struct MsgBuf *msgbuf_p, struct Client *client_p, struct Client *source_p, int parc, const char *parv[])
+{
+	if (!HasPrivilege(source_p, "oper:shedding"))
+	{
+		sendto_one(source_p, form_str(ERR_NOPRIVS), me.name, source_p->name, "SHEDDING");
+		return;
+	}
+
+	if (parc != 4 && !(parc == 3 && irccmp(parv[2], "OFF") == 0))
+	{
+		sendto_one(source_p, form_str(ERR_NEEDMOREPARAMS),
+			me.name, source_p->name, "SHEDDING");
+		return;
+	}
+
+	if (hunt_server(client_p, source_p, ":%s SHEDDING %s %s %s :%s", 1,
+		parc, parv) != HUNTED_ISME)
+		return;
+
+	if (!irccmp(parv[2], "OFF"))
+	{
+		sendto_realops_snomask(SNO_GENERAL, L_ALL | L_NETWIDE, "%s disabled user shedding", get_oper_name(source_p));
+		rb_event_delete(user_shedding_main_ev);
+		user_shedding_main_ev = NULL;
+		rb_event_delete(user_shedding_shed_ev);
+		user_shedding_shed_ev = NULL;
+		return;
+	}
+
+	rate = atoi(parv[2]);
+
+	if(rate < SHED_RATE_MIN)
+	{
+		sendto_one_notice(source_p, "Shedding rate must be at least %d", SHED_RATE_MIN);
+		return;
+	}
+
+	sendto_realops_snomask(SNO_GENERAL, L_ALL | L_NETWIDE, "%s enabled user shedding (interval: %d seconds, reason: %s)",
+		get_oper_name(source_p), rate, parv[3]);
+
+	rate -= (rate/5);
+	rb_event_delete(user_shedding_main_ev);
+	user_shedding_main_ev = NULL;
+	user_shedding_main_ev = rb_event_add("user shedding main event", user_shedding_main, NULL, rate);
+}
+
+static void
+user_shedding_main(void *unused)
+{
+	int deviation = (rate / (3+(int) (7.0f*rand()/(RAND_MAX+1.0f))));
+
+	user_shedding_shed_ev = rb_event_addish("user shedding shed event", user_shedding_shed, NULL, rate+deviation);
+}
+
+static void
+user_shedding_shed(void *unused)
+{
+	rb_dlink_node *ptr;
+	struct Client *client_p;
+
+	RB_DLINK_FOREACH_PREV(ptr, lclient_list.tail)
+	{
+		client_p = ptr->data;
+
+		if (!MyClient(client_p)) /* It could be servers */
```

Comment:
Perhaps `IsClient` or `IsPerson`, since the client must be local because it's on `lclient_list`.